sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala

```diff
@@ -208,10 +208,9 @@ final class SpecificInternalRow(val values: Array[MutableValue]) extends BaseGen
   def this(dataTypes: Seq[DataType]) = {
     // SPARK-32550: use while loop instead of map
     this(new Array[MutableValue](dataTypes.length))
-    val length = values.length
     var i = 0
-    while (i < length) {
-      values(i) = dataTypeToMutableValue(dataTypes(i))
+    dataTypes.foreach { dt =>
+      values(i) = dataTypeToMutableValue(dt)
```

Comment:
Because `schema.fields` is a `Array` , support random access, so there was no change of `def this(schema: StructType) ` constructor.